### PR TITLE
Make dotProd and Backpropagation as Matrix multiplication

### PR DESCRIPTION
**Pull request by diondermaku**

### CODE REVIEW

File: modules/DNN/NeuronLayerMat.h [hunk 1]
```diff
@@ -28,6 +29,7 @@ namespace DirectGraphicalModels {
 			DllExport void	setValues(const Mat& values);
 			DllExport Mat   getValues(void) const { return m_values; }
 			DllExport int   getNumNeurons(void) const { return m_values.rows; }
+            DllExport Mat   getWeights(void) const { return m_weights; }
```
Comment (ereator) [inline modules/DNN/NeuronLayerMat.h:32]: Please roll back all the changes in this file

File: modules/DNN/NeuronLayerMat.cpp [hunk 2]
```diff
@@ -26,12 +26,13 @@ namespace DirectGraphicalModels { namespace dnn
 
 	void CNeuronLayerMat::dotProd(const CNeuronLayerMat& layer)
 	{
-		for (int y = 0; y < m_values.rows; y++) {
-			float value = 0;
-			for (int x = 0; x < layer.m_values.rows; x++)
-				value += layer.m_weights.at<float>(x, y) * layer.m_values.at<float>(x, 0);
-			m_values.at<float>(y, 0) = sigmoidFunction(value);
-		}
+        Mat AA = layer.getValues();
```
Comment (ereator) [inline modules/DNN/NeuronLayerMat.cpp:29]: Please use `layer.m_values`, no additional method is needed here

File: modules/DNN/NeuronLayerMat.cpp [hunk 3]
```diff
@@ -26,12 +26,13 @@ namespace DirectGraphicalModels { namespace dnn
...
+        Mat BB = layer.getWeights();
```
Comment (ereator) [inline modules/DNN/NeuronLayerMat.cpp:30]: Please use `layer.m_weights` here. No call of additional methods is needed.

File: modules/DNN/NeuronLayerMat.cpp [hunk 4]
```diff
@@ -26,12 +26,13 @@ namespace DirectGraphicalModels { namespace dnn
...
+        Mat res = BB.t() * AA;
```
Comment (ereator) [inline modules/DNN/NeuronLayerMat.cpp:31]: Please use `gemm()` method here. This is needed for further optimization.

File: modules/DNN/NeuronLayerMat.cpp [hunk 5]
```diff
@@ -44,23 +45,37 @@ namespace DirectGraphicalModels { namespace dnn
...
-
-		for (int i = 0; i < layerA.getNumNeurons(); i++) 
-			for (int j = 0; j < layerB.getNumNeurons(); j++) 
-				layerA.m_weights.at<float>(i, j) += learningRate * DeltaJ.at<float>(j, 0) * layerA.m_values.at<float>(i, 0);;
+        //  Mat layerB_weights = layerB.getWeights(); // 60 x 10
+        //  Mat layerB_values = layerB.getValues();   // 60 x 1
+        //  Mat layerA_weights = layerA.getValues();  // 784 x 1
 
-		layerB.m_weights += DeltaWjk;
+        Mat DeltaWjk(layerB.getNumNeurons(), layerC.getNumNeurons(), CV_32FC1);
```
Comment (ereator) [inline modules/DNN/NeuronLayerMat.cpp:52]: What about matrix multiplication here? I.e.using `gemm()` instead of the loops?

**Review by ereator — CHANGES_REQUESTED** [review]

File: modules/DNN/NeuronLayerMat.h (hunk 1 above)
Comment (diondermaku) [reply modules/DNN/NeuronLayerMat.h]: done

**Pull request merged**